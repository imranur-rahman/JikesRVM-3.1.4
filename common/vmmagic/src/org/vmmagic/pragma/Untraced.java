/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;

/**
 * Use this annotation to mark fields that are read without the knowledge of the
 * memory management system. This means that barriers are not triggered and
 * the reference is also not traced by the garbage collector.<p>
 *
 * In order to guarantee that the constraints described above are honored by the compiler,
 * all accesses to untraced fields must be made when the field is resolved. If the field
 * were not resolved at the time when the code for the access was being compiled, the
 * Untraced annotation wouldn't be available (annotations are loaded at resolution
 * time).
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
public @interface Untraced { }
