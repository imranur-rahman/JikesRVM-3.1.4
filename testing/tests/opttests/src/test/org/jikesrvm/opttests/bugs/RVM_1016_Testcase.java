/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.opttests.bugs;

import org.vmmagic.pragma.NoInline;

/**
 * Testcase for emitting of ORC instructions on PPC.<p>
 *
 * Incorrect templates for the ORC and ORCr operators led to invalid
 * instructions being emitted. This test case tests only ORC.
 */
public class RVM_1016_Testcase {

  public static void main(String[] args) {
    orc(0, Integer.MIN_VALUE);
    orc(0, Integer.MAX_VALUE);

    orc(Integer.MIN_VALUE, -1);
    orc(Integer.MAX_VALUE, -1);

    orc(0, 0);
    orc(-1, -1);
    orc(-1, 0);
    orc(1, 0);
    orc(0, 1);
    orc(0, -1);
    orc(0, -2);

    orc(0, 2);

    orc(3, 5);
    orc(5, 3);

    orc(2, 1);

    orc(7, 13);
    orc(13, 7);
  }

  @NoInline // prevent removal of arithmetic operations by optimization
  private static void orc(int firstOp, int secondOp) {
    printOperation(firstOp, secondOp);
    int result = firstOp | ~secondOp;
    System.out.println(result);
  }

  @NoInline // inlining this will make the error disappear
  private static void printOperation(int firstOp, int secondOp) {
    System.out.print(firstOp + " | ~ " + secondOp + " : ");
  }

}
