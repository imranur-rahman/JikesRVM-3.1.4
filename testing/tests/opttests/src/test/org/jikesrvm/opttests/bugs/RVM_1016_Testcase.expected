0 | ~ -2147483648 : 2147483647
0 | ~ 2147483647 : -2147483648
-2147483648 | ~ -1 : -2147483648
2147483647 | ~ -1 : 2147483647
0 | ~ 0 : -1
-1 | ~ -1 : -1
-1 | ~ 0 : -1
1 | ~ 0 : -1
0 | ~ 1 : -2
0 | ~ -1 : 0
0 | ~ -2 : 1
0 | ~ 2 : -3
3 | ~ 5 : -5
5 | ~ 3 : -3
2 | ~ 1 : -2
7 | ~ 13 : -9
13 | ~ 7 : -3
